File: tensorflow/tools/api/tests/module_test.py
# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# ==============================================================================
"""Smoke tests for tensorflow module."""

import pkgutil

import tensorflow as tf

from tensorflow.python import tf2
from tensorflow.python.platform import test


class ModuleTest(test.TestCase):

  def testCanLoadWithPkgutil(self):
    out = pkgutil.find_loader('tensorflow')
    self.assertIsNotNone(out)

  def testDocString(self):
    self.assertIn('TensorFlow', tf.__doc__)
    self.assertNotIn('Wrapper', tf.__doc__)

  def testDict(self):
    # Check that a few modules are in __dict__.
    # pylint: disable=pointless-statement
    tf.nn
    tf.keras
    tf.image
    # pylint: enable=pointless-statement
    self.assertIn('nn', tf.__dict__)
    self.assertIn('keras', tf.__dict__)
    self.assertIn('image', tf.__dict__)

  def testName(self):
    self.assertEqual('tensorflow', tf.__name__)

  def testBuiltInName(self):
    # range is a built-in name in Python. Just checking that
    # tf.range works fine.
    if tf2.enabled():
      self.assertEqual(
          'tf.Tensor([1 2 3 4 5 6 7 8 9], shape=(9,), dtype=int32)',
          str(tf.range(1, 10)))
    else:
      self.assertEqual('Tensor("range:0", shape=(9,), dtype=int32)',
                       str(tf.range(1, 10)))

  def testCompatV2HasCompatV1(self):
    # pylint: disable=pointless-statement
    tf.compat.v2.compat.v1.keras
    # pylint: enable=pointless-statement

  def testSummaryMerged(self):
    # pylint: disable=pointless-statement
    tf.summary.image
    # If we use v2 API, check for create_file_writer,
    # otherwise check for FileWriter.
    if hasattr(tf, '_major_api_version') and tf._major_api_version == 2:
      tf.summary.create_file_writer
    else:
      tf.compat.v1.summary.FileWriter
    # pylint: enable=pointless-statement

  def testPythonModuleIsHidden(self):
    self.assertNotIn('python', dir(tf))


if __name__ == '__main__':
  test.main()

File: tensorflow/compiler/xla/runtime/module_test.cc
/* Copyright 2022 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow/compiler/xla/runtime/module.h"

#include <memory>
#include <optional>

#include "tensorflow/compiler/xla/runtime/custom_call.h"
#include "tensorflow/tsl/platform/test.h"

namespace xla {
namespace runtime {

struct TestStateRef : public Module::StateRef {};

struct TestState : public Module::State {
  explicit TestState(int32_t value) : value(value) {}

  int32_t value;
};

struct ModuleNoStateRef : public StatefulModule<TestState> {
  using Base = StatefulModule<TestState>;

  ModuleNoStateRef() : Base("module-no-state-ref") {}

  absl::StatusOr<std::unique_ptr<TestState>> CreateModuleState() const final {
    return std::make_unique<TestState>(42);
  }
};

struct ModuleWithStateRef : public StatefulModule<TestState, TestStateRef> {
  using Base = StatefulModule<TestState, TestStateRef>;

  ModuleWithStateRef() : Base("module-with-state-ref") {}

  absl::StatusOr<std::unique_ptr<TestState>> CreateModuleState() const final {
    return std::make_unique<TestState>(42);
  }
};

struct ModuleNoState : public StatelessModule {
  ModuleNoState() : StatelessModule("module-no-state") {}
};

TEST(ModuleTest, ModuleNoStateRef) {
  ModuleNoStateRef module;
  EXPECT_EQ(module.name(), "module-no-state-ref");

  auto state = module.CreateModuleState();
  ASSERT_TRUE(state.ok());
  EXPECT_EQ((*state)->value, 42);

  CustomCall::UserData user_data;
  ASSERT_TRUE(module.InitializeUserData(state->get(), user_data).ok());
}

TEST(ModuleTest, ModuleWithStateRef) {
  ModuleWithStateRef module;
  EXPECT_EQ(module.name(), "module-with-state-ref");

  auto state = module.CreateModuleState();
  ASSERT_TRUE(state.ok());
  EXPECT_EQ((*state)->value, 42);

  CustomCall::UserData user_data;
  ASSERT_TRUE(module.InitializeUserData(state->get(), user_data).ok());
}

TEST(ModuleTest, ModuleNoState) {
  ModuleNoState module;
  EXPECT_EQ(module.name(), "module-no-state");

  auto state = dynamic_cast<Module&>(module).CreateState();
  ASSERT_TRUE(state.ok());
  EXPECT_FALSE(state->get());

  CustomCall::UserData user_data;
  ASSERT_TRUE(module.InitializeUserData(user_data).ok());
}

}  // namespace runtime
}  // namespace xla
